src/uct/ib/base/ib_device.c

```diff
@@ -983,6 +983,76 @@ ucs_status_t uct_ib_device_port_check(uct_ib_device_t *dev, uint8_t port_num,
     return UCS_OK;
 }
 
+ucs_status_t
+uct_ib_device_ece_check(uct_ib_device_t *dev,
+                        struct ibv_context *ctx, struct ibv_pd *pd)
+{
+    ucs_status_t status = UCS_OK;
+#ifdef HAVE_DECL_IBV_ECE
+    struct ibv_ece ece  = {};
+    struct ibv_qp *dummy_qp;
+    struct ibv_cq *cq;
+    struct ibv_qp_init_attr qp_init_attr;
+
+    cq = ibv_create_cq(ctx, 1, NULL, NULL, 0);
+    if (cq == NULL) {
+        status = UCS_ERR_IO_ERROR;
+        goto out;
+    }
+
+    memset(&qp_init_attr, 0, sizeof(qp_init_attr));
+    qp_init_attr.send_cq = cq;
+    qp_init_attr.recv_cq = cq;
+    qp_init_attr.qp_type = IBV_QPT_RC;
+
+    qp_init_attr.cap.max_send_wr  = 1;
+    qp_init_attr.cap.max_recv_wr  = 1;
+    qp_init_attr.cap.max_send_sge = 1;
+    qp_init_attr.cap.max_recv_sge = 1;
+
+    dummy_qp = ibv_create_qp(pd, &qp_init_attr);
+    if (dummy_qp == NULL) {
+        status = UCS_ERR_IO_ERROR;
+        goto free_cq;
+    }
+
+    /* ibv_set_ece check whether ECE is supported */
+    if ((ibv_query_ece(dummy_qp, &ece) == 0) &&
+        (ibv_set_ece(dummy_qp, &ece) == 0)) {
+        dev->flags |= UCT_IB_DEVICE_FLAG_ECE;
+    }
+
+    ibv_destroy_qp(dummy_qp);
+free_cq:
+    ibv_destroy_cq(cq);
+out:
+#endif
+    return status;
+}
+
+ucs_status_t
+uct_ib_set_ece(uct_ib_device_t *dev, struct ibv_qp *qp, uint32_t ece_val)
+{
+    struct ibv_ece UCS_V_UNUSED ece;
+
+    if (ece_val == 0) {
+        return UCS_OK;
+    }
+
+    if ((dev->flags & UCT_IB_DEVICE_FLAG_ECE) == 0) {
+        return UCS_ERR_UNSUPPORTED;
+    }
```

Comment:
since uct_rc_iface_t already checks UCT_IB_DEVICE_FLAG_ECE and should not allow using ece!=0 if that flag is not set, this should be converted to an assertion
`ucs_assertv_always(dev->flags & UCT_IB_DEVICE_FLAG_ECE, "device=%s ece=0x%x", device_name, ece_val);`

Reply:
Done